File: fixtures/sample-repos/playwright/pages/actions/LoginActionsBad.ts
import type { Page } from '@playwright/test';
import { expect } from '@playwright/test';
import { LoginLocators } from '../locators/LoginLocators';

// VIOLATION FILE: expect-in-action — this action class contains an
// `expect(...)` call. Specs own assertions; actions return state or
// perform interactions. The canonical pattern is in LoginActions.ts
// (no `expect`, no `assert`); this file exists to demonstrate the rule.
export class LoginActionsBad {
  locators: LoginLocators;

  constructor(private page: Page) {
    this.locators = new LoginLocators(page);
  }

  async signIn(email: string, password: string): Promise<void> {
    await this.locators.emailInput.fill(email);
    await this.locators.passwordInput.fill(password);
    // VIOLATION: expect/assert APIs belong in spec files, not action classes.
    expect(email).toContain('@');
    await this.locators.signInButton.click();
  }
}

File: fixtures/sample-repos/playwright-dotnet/Pages/Actions/LoginActionsBad.cs
using Microsoft.Playwright;
using Sample.PlaywrightDotnet.Pages.Locators;

namespace Sample.PlaywrightDotnet.Pages.Actions;

// VIOLATION FILE: Expect in action class — specs own assertions; actions perform
// interactions only. Canonical pattern is LoginActions.cs (no Expect/Assert).
public class LoginActionsBad
{
    private readonly LoginLocators _locators;

    public LoginActionsBad(IPage page)
    {
        _locators = LoginLocators.For(page);
    }

    public async Task SignInAsync(string email, string password)
    {
        await _locators.EmailInput.FillAsync(email);
        await _locators.PasswordInput.FillAsync(password);
        // VIOLATION: assertion APIs belong in test files, not action classes.
        await Expect(_locators.EmailInput).ToBeEditableAsync();
        await _locators.SignInButton.ClickAsync();
    }
}
